src/posix/platform/infra_if.cpp

```diff
@@ -0,0 +1,355 @@
+    }
+
+    rval = setsockopt(sock, IPPROTO_IPV6, IPV6_MULTICAST_HOPS, &kHopLimit, sizeof(kHopLimit));
+    if (rval < 0)
+    {
+        otLogCritPlat("Can't set IPV6_MULTICAST_HOPS: %s", strerror(errno));
+        DieNow(OT_EXIT_ERROR_ERRNO);
+    }
+
+    sInfraIfIcmp6Socket = sock;
+    InitLinkLocalAddress();
+}
+
+void platformInfraIfDeinit(void)
+{
+    if (sInfraIfIcmp6Socket != -1)
+    {
+        close(sInfraIfIcmp6Socket);
+        sInfraIfIcmp6Socket = -1;
+    }
+
+    sInfraIfIndex = 0;
+}
+
+void platformInfraIfUpdateFdSet(fd_set &aReadFdSet, int &aMaxFd)
+{
+    VerifyOrExit(sInfraIfIcmp6Socket != -1);
+
+    FD_SET(sInfraIfIcmp6Socket, &aReadFdSet);
+    aMaxFd = OT_MAX(aMaxFd, sInfraIfIcmp6Socket);
+
+exit:
+    return;
+}
+
+void platformInfraIfProcess(otInstance *aInstance, const fd_set &aReadFdSet)
+{
+    otError  error = OT_ERROR_DROP;
+    uint8_t  buffer[1500];
+    uint16_t bufferLength;
+
+    ssize_t         rval;
+    struct msghdr   msg;
+    struct iovec    bufp;
+    char            cmsgbuf[128];
+    struct cmsghdr *cmh;
+    uint32_t        ifIndex  = 0;
+    int             hopLimit = -1;
+
+    struct sockaddr_in6 srcAddr;
+    struct in6_addr     dstAddr;
+
+    VerifyOrExit(sInfraIfIcmp6Socket != -1);
+    VerifyOrExit(FD_ISSET(sInfraIfIcmp6Socket, &aReadFdSet));
+
+    memset(&srcAddr, 0, sizeof(srcAddr));
+    memset(&dstAddr, 0, sizeof(dstAddr));
+
+    bufp.iov_base      = buffer;
+    bufp.iov_len       = sizeof(buffer);
+    msg.msg_iov        = &bufp;
+    msg.msg_iovlen     = 1;
+    msg.msg_name       = &srcAddr;
+    msg.msg_namelen    = sizeof(srcAddr);
+    msg.msg_control    = cmsgbuf;
+    msg.msg_controllen = sizeof(cmsgbuf);
+
+    rval = recvmsg(sInfraIfIcmp6Socket, &msg, 0);
+    if (rval < 0)
+    {
+        otLogWarnPlat("failed to receive ICMPv6 message: %s", strerror(errno));
+        ExitNow();
+    }
+
+    bufferLength = static_cast<uint16_t>(rval);
+
+    for (cmh = CMSG_FIRSTHDR(&msg); cmh; cmh = CMSG_NXTHDR(&msg, cmh))
+    {
+        if (cmh->cmsg_level == IPPROTO_IPV6 && cmh->cmsg_type == IPV6_PKTINFO &&
+            cmh->cmsg_len == CMSG_LEN(sizeof(struct in6_pktinfo)))
+        {
+            struct in6_pktinfo pktinfo;
+
+            memcpy(&pktinfo, CMSG_DATA(cmh), sizeof pktinfo);
+            ifIndex = pktinfo.ipi6_ifindex;
+            dstAddr = pktinfo.ipi6_addr;
+        }
+        else if (cmh->cmsg_level == IPPROTO_IPV6 && cmh->cmsg_type == IPV6_HOPLIMIT &&
+                 cmh->cmsg_len == CMSG_LEN(sizeof(int)))
+        {
+            hopLimit = *(int *)CMSG_DATA(cmh);
+        }
+    }
+
+    VerifyOrExit(ifIndex == sInfraIfIndex);
+
+    // We currently acepts only RA & RS messages for the Border Router and it requires that
+    // the hoplimit must be 255.
+    VerifyOrExit(hopLimit == 255);
+
+    // Drop multicast messages sent by ourselves.
+    VerifyOrExit(!otIp6IsAddressEqual(&sInfraIfLinkLocalAddr, reinterpret_cast<otIp6Address *>(&srcAddr.sin6_addr)));
```

Comment:
We rely on multicast loop to send RA to the infra link to generate SLAAC address. We drop the message here just for simplicity (so that the routing manager will not receive the PIO, RIO sent by itself).

We will remove this in our enhance PR.